### PR TITLE
More logging for Raft-based topology

### PR DESCRIPTION
**Pull request by kbr-scylla**
Currently if topology coordinator gets stuck in a CI test run it's hard to debug this (e.g. scylladb/scylladb#16708). We can add a lot of logging inside topology coordinator code to aid debugging, without spamming the logs -- these are relatively rare control plane events.

### CODE REVIEW

**Comment by kbr-scylla** [comment]
v2: adjusted tests

**Comment by kbr-scylla** [comment]
@benipeled looks like the CI job hanged at the end https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5734/console

**Comment by benipeled** [comment]
> @benipeled looks like the CI job hanged at the end https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5734/console

The CI "fell" on timeout due to `boost.sstable_compaction_test.max_ongoing_compaction_test` running without completion. The logfile is 24GB (you went too far with the "More logging" :) so it's a bit hard to parse it, but it's full of
```
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - No compaction necessary for bucket size 1 , key 1705413600000000, now 1705413600000000
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - [0x60b000014c68] Going to compact 0 non-expired sstables
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170001e4200 for table tests.8 compaction_group=0/1 [0x60300006a0f0]: sstables=0 can_proceed=true auto_compaction=false
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x61700057b500 for table tests.9 compaction_group=0/1 [0x60300006c1c0]: done
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x61700045d300 for table tests.8 compaction_group=0/1 [0x60300006a0f0]: switch_state: pending -> none: pending=3 active=0 done=28 errors=0
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170001e4200 for table tests.8 compaction_group=0/1 [0x60300006a0f0]: done
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x61700049e280 for table tests.9 compaction_group=0/1 [0x60300006c1c0]: switch_state: pending -> none: pending=2 active=0 done=28 errors=0
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170001e4200 for table tests.8 compaction_group=0/1 [0x60300006a0f0]: switch_state: pending -> none: pending=1 active=0 done=28 errors=0
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x61700057b500 for table tests.9 compaction_group=0/1 [0x60300006c1c0]: switch_state: pending -> none: pending=0 active=0 done=28 errors=0
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170003a5700 for table tests.8 compaction_group=0/1 [0x60300006a0f0]: started
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170003ab200 for table tests.9 compaction_group=0/1 [0x60300006c1c0]: started
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170003a5700 for table tests.8 compaction_group=0/1 [0x60300006a0f0]: switch_state: none -> pending: pending=1 active=0 done=28 errors=0
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170003ab200 for table tests.9 compaction_group=0/1 [0x60300006c1c0]: switch_state: none -> pending: pending=2 active=0 done=28 errors=0
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - [0x60b000014c68] TWCS skipping check for fully expired SSTables
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - time_window_compaction_strategy::newest_bucket:
  now 1705413600000000
  buckets = {
    key=1705413600000000, size=1
  }

DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - No compaction necessary for bucket size 1 , key 1705413600000000, now 1705413600000000
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - [0x60b000014c68] Going to compact 0 non-expired sstables
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction_manager - Compaction task 0x6170003a5700 for table tests.8 compaction_group=0/1 [0x60300006a0f0]: sstables=0 can_proceed=true auto_compaction=false
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - [0x60b0000154a8] TWCS skipping check for fully expired SSTables
DEBUG 2024-01-12 14:00:07,999 [shard 0:main] compaction - time_window_compaction_strategy::newest_bucket:
  now 1705413600000000
  buckets = {
    key=1705413600000000, size=1
  }
````

And later I see a few
```
DEBUG 2024-01-12 14:59:57,045 [shard 0:main] compaction_manager - resubmitting postponed compaction for table tests.0 compaction_group=table_for_tests::table_state [0x613000011d80]
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x6170003da980 for table tests.0 compaction_group=table_for_tests::table_state [0x613000011d80]: started
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x617000234d80 for table tests.6 compaction_group=table_for_tests::table_state [0x613000013d00]: switch_state: none -> pending: pending=7 active=0 done=30 errors=0
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x6170002b8880 for table tests.9 compaction_group=0/1 [0x60300006c1c0]: done
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x6170001b1600 for table tests.7 compaction_group=0/1 [0x603000068020]: sstables=0 can_proceed=true auto_compaction=false
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x61700042b500 for table tests.8 compaction_group=table_for_tests::table_state [0x613000014780]: sstables=0 can_proceed=true auto_compaction=false
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - resubmitting postponed compaction for table tests.1 compaction_group=table_for_tests::table_state [0x6130000122c0]
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x6170001b1600 for table tests.7 compaction_group=0/1 [0x603000068020]: done
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - resubmitting postponed compaction for table tests.4 compaction_group=0/1 [0x603000061e70]
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - resubmitting postponed compaction for table tests.8 compaction_group=0/1 [0x60300006a0f0]
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x6170005d3080 for table tests.1 compaction_group=table_for_tests::table_state [0x6130000122c0]: started
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x6170003da980 for table tests.0 compaction_group=table_for_tests::table_state [0x613000011d80]: switch_state: none -> pending: pending=8 active=0 done=30 errors=0
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - Compaction task 0x61700042b500 for table tests.8 compaction_group=table_for_tests::table_state [0x613000014780]: done
DEBUG 2024-01-12 14:59:57,046 [shard 0:main] compaction_manager - resubmitting postponed compaction for table tests.1 compaction_group=0/1 [0x60300005c4d0]
```
Which for me looks normal but you might see something unusual

**Comment by mykaul** [comment]
> The CI "fell" on timeout due to `boost.sstable_compaction_test.max_ongoing_compaction_test` running without completion. The logfile is 24GB (you went too far with the "More logging" :) so it's a bit hard to parse it, but it's full of

The fact we do not log rotate is what caused a 24GB file. We must implement it, sooner rather than later.

**Comment by benipeled** [comment]
> > The CI "fell" on timeout due to `boost.sstable_compaction_test.max_ongoing_compaction_test` running without completion. The logfile is 24GB (you went too far with the "More logging" :) so it's a bit hard to parse it, but it's full of
> 
> The fact we do not log rotate is what caused a 24GB file. We must implement it, sooner rather than later.

The only change with logrotate (if it's possible to implement while ./test.py is writing) was the need to search the root-cause in multiple small files instead of one large file - the problem isn't to load the file but to search for a needle in 24G logs

**Comment by gleb-cloudius** [comment]
And again lets push this one down the road since it will cause huge rebase for my two outstanding series and probably @gusev-p one that they are waiting for.

**Comment by kbr-scylla** [comment]
> you went too far with the "More logging"

I didn't add any logging to compactions. Here we see what looks like a hot loop in compactions spamming the logs.

@denesb please take a look or assign someone -- producing 24GB compaction logs in CI is a bit too much.

**Comment by kbr-scylla** [comment]
cc @raphaelsc -- I see you added debug level compaction logging in 63e4d6c96565344ab68a82d8306317a976069372
perhaps the hot loop -> spam encountered in the CI run above is due to some issue you tried to debug?

**Comment by kbr-scylla** [comment]
The commit mentions https://github.com/scylladb/scylladb/issues/13280 where the same test would've timed out. So I think it's possible we hit the same issue

**Comment by kbr-scylla** [comment]
> And again lets push this one down the road since it will cause huge rebase for my two outstanding series and probably @gusev-p one that they are waiting for.

@gleb-cloudius can we push this now?

**Comment by gleb-cloudius** [comment]
> > And again lets push this one down the road since it will cause huge rebase for my two outstanding series and probably @gusev-p one that they are waiting for.
> 
> @gleb-cloudius can we push this now?

OK. I hope the conflict with request tracking will not be huge.

**Comment by kbr-scylla** [comment]
Rebased

File: service/storage_service.cc [hunk 1]
```diff
@@ -2374,6 +2374,7 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
 
         switch (*tstate) {
             case topology::transition_state::join_group0: {
+                rtlogger.info("entered `join_group0` transition state");
```
Comment (gleb-cloudius) [inline service/storage_service.cc:2377]: Why can't you log it once before the switch:  `rtlogger.info("entered {} transition state", topology::transition_state::join_group0);`

File: test/pylib/scylla_cluster.py [hunk 2]
```diff
@@ -127,7 +127,8 @@ def make_scylla_conf(workdir: pathlib.Path, host_addr: str, seed_addrs: List[str
     '--abort-on-lsa-bad-alloc', '1',
     '--abort-on-seastar-bad-alloc',
     '--abort-on-internal-error', '1',
-    '--abort-on-ebadf', '1'
+    '--abort-on-ebadf', '1',
+    '--logger-log-level', 'raft_topology=debug',
```
Comment (gleb-cloudius) [inline test/pylib/scylla_cluster.py:131]: Shouldn't we do it for topology suits only?

Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:131]: Doesn't hurt to do it for others as well.

File: service/storage_service.cc (hunk 1 above)
Comment (kbr-scylla) [reply service/storage_service.cc]: Right, fixed

**Comment by kbr-scylla** [comment]
- rebased
- one logging call for transition_state before the `switch`

**Review by gleb-cloudius — APPROVED** [review]

**Comment by kbr-scylla** [comment]
@scylladb/scylla-maint please review/merge

**Comment by kbr-scylla** [comment]
@scylladb/scylla-maint ping. This is a simple PR, although boilerplate'y. Best to merge it quickly as it's easy to conflict with it.

**Comment by kbr-scylla** [comment]
@scylladb/scylla-maint ping, please review/merge
